File: deploy.go
package main

import (
	"log"
	"os"
	"os/exec"
	"strings"

	"github.com/joho/godotenv"
)

func main() {
	const example_cmd = "go run deploy <network> <part> <simulation=true>"

	// Source .env
	err := godotenv.Load()
	if err != nil {
		log.Fatal("Error loading .env file")
	}

	// Validate private key
	privateKey := os.Getenv("PRIVATE_KEY")
	if privateKey == "" {
		log.Fatal("PRIVATE_KEY not set in .env")
	}
	if privateKey[:2] != "0x" {
		log.Fatal("PRIVATE_KEY must start with 0x")
	}

	// Require forge
	forgePath, err := exec.LookPath("forge")
	if err != nil {
		log.Fatal("Forge not found. Please install foundry.")
	}

	// Require network & setup args
	args := os.Args[1:]
	if len(args) < 1 {
		log.Fatal("Incorrect number of arguments. Example: \n", example_cmd)
	}
	network := args[0]
	part := 1
	if len(args) > 1 {
		switch args[1] {
		case "1":
			part = 1
		case "2":
			part = 2
		case "3":
			part = 3
		default:
			log.Fatal("Invalid part number. Example: \n", example_cmd)
		}
	}
	isSim := true
	if len(args) > 2 {
		isSim = args[2] != "false"
	}

	// Get RPC from env & ensure it exists
	envPath := "ETH_NODE_URI_" + strings.ToUpper(network)
	rpc := os.Getenv(envPath)
	if rpc == "" {
		log.Fatal("RPC not found for network ("+envPath+"): ", network)
	}

	// Build & run command
	dir, _ := os.Getwd()
	var scriptFile string
	switch part {
	case 2:
		scriptFile = "/script/DeployCurvance2.s.sol"
	case 3:
		scriptFile = "/script/DeployCurvance3.s.sol"
	default:
		scriptFile = "/script/DeployCurvance.s.sol"
	}

	forgeArgs := []string{"script", dir + scriptFile, network, "--sig", "run(string)", "--rpc-url", rpc, "-vvvv", "--ffi"}
	if isSim {
		log.Printf("Deploying to %s [TEST-RUN]\n", network)
	} else {
		forgeArgs = append(forgeArgs, "--broadcast", "--skip-simulation", "--priority-gas-price", "15")
		log.Printf("Deploying to %s\n", network)
		log.Println("REMEMBER: UPDATE INDEXER & DAPP WITH NEW CONTRACT ADDRESS")
		log.Println("REMEMBER: UPDATE INDEXER & DAPP WITH NEW CONTRACT ADDRESS")
		log.Println("REMEMBER: UPDATE INDEXER & DAPP WITH NEW CONTRACT ADDRESS")
	}
	cmd := exec.Command(forgePath, forgeArgs...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	cmd.Run()
}

File: deploy.js
const { execSync } = require('child_process');
const ethers = require('ethers');
require('dotenv').config();

const exampleCmd = 'node deploy.js <network> <simulation=true> <resume=false>';
let retry_count = 0;
main().catch(console.error);

async function main() {
  const { isSim, network, rpc, resume } = setup();
  const scripts = [
    'script/DeployTestnetTokens.s.sol',
    'script/DeploySetupOracles.s.sol',
    'script/DeployExecuteRedstoneFeeds.s.sol',
    'script/DeployCurvance.s.sol'
  ];

  const generateForgeScript = (script) => {
    // Build & run command
    let forgeArgs = [
      script,
      `"${network}" --sig "run(string)"`,
      `--rpc-url ${rpc}`,
      '-vvvv',
      '--ffi',
      `--with-gas-price 30gwei`,
    ];

    if (resume) {
      forgeArgs.push('--resume');
    }

    if (isSim) {
      console.log(`Deploying to ${network} [TEST-RUN]`);
    } else {
      forgeArgs = forgeArgs.concat([
        '--broadcast',
        '--skip-simulation',
        '--no-storage-caching',
        '--priority-gas-price 200',
        '-g 230'
      ]);
      console.log(`Deploying to ${network}`);
    }

    return forgeArgs;
  }

  let finished_deploy = false;
  try {
    for (const script of scripts) {
      const forgeArgs = generateForgeScript(script);
      console.log(`Running forge script for ${script}`);
      execSync(`forge script ${forgeArgs.join(' ')}`, { stdio: 'inherit' });
    }

    finished_deploy = true;
  } catch (err) {
    console.error('\nError executing forge command.');
  }

  // Parse broadcast results to see what deployments made it
  if (!isSim) {
    console.log('REMEMBER: UPDATE INDEXER & DAPP WITH NEW CONTRACT ADDRESS');
    console.log('REMEMBER: UPDATE INDEXER & DAPP WITH NEW CONTRACT ADDRESS');
    console.log('REMEMBER: UPDATE INDEXER & DAPP WITH NEW CONTRACT ADDRESS');

    if (!finished_deploy && resume) {
      retry_count++
      console.log(`Deployment failed (${retry_count}), retrying...`);

      await main();
    }
  }
}

function setup() {
  // Load .env file
  if (!process.env.PRIVATE_KEY) {
    console.error('Error loading .env file');
    process.exit(1);
  }

  // Validate private key
  const privateKey = process.env.PRIVATE_KEY;
  if (!privateKey) {
    console.error('PRIVATE_KEY not set in .env');
    process.exit(1);
  }
  if (!privateKey.startsWith('0x')) {
    console.error('PRIVATE_KEY must start with 0x');
    process.exit(1);
  }

  // Require forge
  let forgePath;
  try {
    forgePath = execSync('which forge').toString().trim();
  } catch (err) {
    console.error('Forge not found. Please install foundry.');
    process.exit(1);
  }

  // Require network & setup args
  const args = process.argv.slice(2);
  if (args.length < 1) {
    console.error('Incorrect number of arguments. Example: \n', exampleCmd);
    process.exit(1);
  }
  const network = args[0];
  let isSim = true;
  if (args.length > 1) {
    isSim = args[1] !== 'false';
  }

  let resume = false;
  if (args.length > 2) {
    resume = args[2] !== 'false';
  }

  // Get RPC from env & ensure it exists
  const envPath = `ETH_NODE_URI_${network.toUpperCase()}`;
  const rpc = process.env[envPath];
  if (!rpc) {
    console.error(`RPC not found for network (${envPath}): `, network);
    process.exit(1);
  }
  const provider = new ethers.JsonRpcProvider(rpc);

  return { privateKey, forgePath, isSim, resume, network, rpc, provider };
}
